logic/concludable/ConjunctionConcludable.java

```diff
@@ -112,21 +119,107 @@ public Value asValue() {
+            Variable conjVar, Variable headVar, Map<Reference, Set<Reference>> variableMapping) {
+        if (varHintsDisjoint(headVar, conjVar)) return Optional.empty();
+        Map<Reference, Set<Reference>> cloneOfMapping = cloneMapping(variableMapping);
+        cloneOfMapping.putIfAbsent(conjVar.reference(), new HashSet<>());
+        cloneOfMapping.get(conjVar.reference()).add(headVar.reference());
+        return Optional.of(cloneOfMapping);
+    }
+
+    <T> Map<T, Set<T>> cloneMapping(Map<T, Set<T>> mapping) {
+        Map<T, Set<T>> clone = new HashMap<>();
+        mapping.forEach((key, set) -> clone.put(key, new HashSet<>(set)));
+        return clone;
+    }
+
     public static class Relation extends ConjunctionConcludable<RelationConstraint, ConjunctionConcludable.Relation> {
 
         public Relation(final RelationConstraint constraint) {
             super(copyConstraint(constraint));
         }
 
         @Override
-        public Stream<Unification> unify(ThenConcludable.Relation unifyWith) {
-            // Check the relation variables' isa constraint labels and prune if there is no intersection
-
-            // Find all roleplayer mapping combinations, which should have the form:
-            // Set<List<Pair<RelationConstraint.RolePlayer, RelationConstraint.RolePlayer>>> rolePlayerMappings
-            // For each, prune if there is no label intersection (or any other pruning, e.g. by value)
-            // Then build a Unification for each valid combination
-            return Stream.empty(); // TODO
+        public Stream<Map<Reference, Set<Reference>>> unify(ThenConcludable.Relation unifyWith) {
+            if (this.constraint().players().size() > unifyWith.constraint().players().size()) return Stream.empty();
+            Map<Reference, Set<Reference>> variableUnifier = new HashMap<>();
+            Optional<Map<Reference, Set<Reference>>> newUnifier;
+            if (constraint.owner().reference().isName()) {
+                newUnifier = extendUnifier(this.constraint().owner(), unifyWith.constraint().owner(), variableUnifier);
+                if (newUnifier.isPresent()) variableUnifier = newUnifier.get();
+                else return Stream.empty();
+            }
+
+            if (constraint.owner().isa().isPresent() && constraint.owner().isa().get().type().reference().isName()) {
+                assert unifyWith.constraint().owner().isa().isPresent();
+                newUnifier = extendUnifier(this.constraint().owner().isa().get().type(),
+                        unifyWith.constraint().owner().isa().get().type(), variableUnifier);
+                if (newUnifier.isPresent()) variableUnifier = newUnifier.get();
+                else return Stream.empty();
+            }
+
+            Map<Reference, Set<Reference>> finalVariableUnifier = variableUnifier;
+            return unifyRolePlayers(this.constraint().players(), unifyWith.constraint().players(),
```

Comment:
can you wrap each of these lists in a `Collections.unmodifableList()` call to signal that they shuoldn't be cut up/modified?

Reply:
Made that change.